File: backend/dashboard_manager.py
"""Python peer of src/managers/dashboardmanager.{h,cpp}.

Both backends scan the same on-disk JSON layout:

- Built-in presets: frontend/dashboards/presets/*.json (read-only)
- User dashboards:  <app_data_dir>/dashboards/*.json  (read/write)

The QML frontend binds to a single `dashboardManager` context property whose
public API must stay identical across backends. Keep this file and the C++
header aligned — if you add a Slot here, add the Q_INVOKABLE on the C++ side
in the same change set.
"""

import json
import os
import re
import tempfile
import time

from PySide6.QtCore import QObject, Property, Signal, Slot

from backend.logging_config import get_logger

logger = get_logger(__name__)


class DashboardManager(QObject):
    dashboardsChanged = Signal()

    def __init__(self, parent=None):
        super().__init__(parent)
        self._presets_dir = ""
        self._user_dir = ""
        self._dashboards = []

    # ------------------------------------------------------------------
    # Configuration
    # ------------------------------------------------------------------

    def setPresetsDir(self, absolute_path: str):
        self._presets_dir = absolute_path or ""

    def setUserDir(self, absolute_path: str):
        self._user_dir = absolute_path or ""
        self._rescan_all()

    # ------------------------------------------------------------------
    # QML-visible API
    # ------------------------------------------------------------------

    @Property('QVariantList', notify=dashboardsChanged)
    def dashboards(self):
        return self._dashboards

    @Slot(str, result='QVariant')
    def loadDashboard(self, dashboard_id: str):
        entry = self._find_entry(dashboard_id)
        if not entry:
            logger.warning("loadDashboard: unknown id %s", dashboard_id)
            return {}
        return self._read_full_spec(entry["path"])

    @Slot(str, result=bool)
    def isBuiltIn(self, dashboard_id: str) -> bool:
        return bool(self._find_entry(dashboard_id).get("builtIn", False))

    @Slot('QVariantMap', result=str)
    def saveDashboard(self, spec) -> str:
        spec = dict(spec or {})
        dashboard_id = str(spec.get("id", "") or "")
        label = str(spec.get("label", "") or "")

        if not dashboard_id or not label:
            logger.warning("saveDashboard requires id and label")
            return ""
        if not self._user_dir:
            logger.warning("User dashboards dir not configured")
            return ""

        existing = self._find_entry(dashboard_id)
        if existing and existing.get("builtIn"):
            logger.warning(
                "Cannot save: id %s collides with a built-in preset",
                dashboard_id,
            )
            return ""

        path = os.path.join(self._user_dir, dashboard_id + ".json")
        if not self._write_spec(path, spec):
            return ""

        self._rescan_all()
        return dashboard_id

    @Slot(str, result=bool)
    def deleteDashboard(self, dashboard_id: str) -> bool:
        entry = self._find_entry(dashboard_id)
        if not entry:
            return False
        if entry.get("builtIn"):
            logger.warning("Refusing to delete built-in dashboard: %s", dashboard_id)
            return False
        try:
            os.remove(entry["path"])
        except OSError as e:
            logger.warning("Failed to delete %s: %s", entry["path"], e)
            return False
        self._rescan_all()
        return True

    @Slot(str, str, result=str)
    def duplicateDashboard(self, source_id: str, new_label: str) -> str:
        entry = self._find_entry(source_id)
        if not entry:
            logger.warning("duplicateDashboard: unknown source id %s", source_id)
            return ""

        spec = self._read_full_spec(entry["path"])
        if not spec:
            return ""

        final_label = new_label if new_label else f"{entry.get('label', '')} (Copy)"
        new_id = self._unique_user_id_from_label(final_label)

        spec["id"] = new_id
        spec["label"] = final_label

        return self.saveDashboard(spec)

    @Slot()
    def refresh(self):
        self._rescan_all()

    # ------------------------------------------------------------------
    # Internals
    # ------------------------------------------------------------------

    def _find_entry(self, dashboard_id: str) -> dict:
        for entry in self._dashboards:
            if entry.get("id") == dashboard_id:
                return entry
        return {}

    def _existing_ids(self) -> set:
        return {entry.get("id", "") for entry in self._dashboards}

    def _ensure_user_dir(self):
        if self._user_dir:
            os.makedirs(self._user_dir, exist_ok=True)

    def _rescan_all(self):
        self._dashboards = []
        taken_ids = set()

        # Built-ins first. Any user JSON that tries to reuse a built-in id
        # gets rejected below (same rule the C++ side enforces).
        if self._presets_dir:
            if os.path.isdir(self._presets_dir):
                for fname in sorted(os.listdir(self._presets_dir)):
                    if not fname.endswith(".json"):
                        continue
                    path = os.path.join(self._presets_dir, fname)
                    header = self._read_header(path)
                    if not header:
                        continue
                    dashboard_id = header.get("id", "")
                    if not dashboard_id or dashboard_id in taken_ids:
                        logger.warning(
                            "Skipping preset with missing or duplicate id: %s",
                            fname,
                        )
                        continue
                    self._dashboards.append({
                        "id": dashboard_id,
                        "label": header.get("label", ""),
                        "builtIn": True,
                        "path": path,
                    })
                    taken_ids.add(dashboard_id)
            else:
                logger.warning("Presets dir does not exist: %s", self._presets_dir)

        # User dashboards.
        self._ensure_user_dir()
        if self._user_dir and os.path.isdir(self._user_dir):
            for fname in sorted(os.listdir(self._user_dir)):
                if not fname.endswith(".json"):
                    continue
                path = os.path.join(self._user_dir, fname)
                header = self._read_header(path)
                if not header:
                    continue
                dashboard_id = header.get("id", "")
                if not dashboard_id:
                    logger.warning(
                        "Skipping user dashboard with missing id: %s", fname,
                    )
                    continue
                if dashboard_id in taken_ids:
                    logger.warning(
                        "Skipping user dashboard — id collides with built-in: "
                        "%s (%s)", dashboard_id, fname,
                    )
                    continue
                self._dashboards.append({
                    "id": dashboard_id,
                    "label": header.get("label", ""),
                    "builtIn": False,
                    "path": path,
                })
                taken_ids.add(dashboard_id)

        logger.info("Scanned dashboards: %d total", len(self._dashboards))
        self.dashboardsChanged.emit()

    def _read_header(self, absolute_path: str) -> dict:
        full = self._read_full_spec(absolute_path)
        if not full:
            return {}
        return {"id": full.get("id", ""), "label": full.get("label", "")}

    def _read_full_spec(self, absolute_path: str) -> dict:
        try:
            with open(absolute_path, "r", encoding="utf-8") as f:
                data = json.load(f)
        except (OSError, json.JSONDecodeError) as e:
            logger.warning("Cannot read dashboard file %s: %s", absolute_path, e)
            return {}
        if not isinstance(data, dict):
            logger.warning(
                "Dashboard JSON root must be an object: %s", absolute_path,
            )
            return {}
        return data

    def _write_spec(self, absolute_path: str, spec: dict) -> bool:
        directory = os.path.dirname(absolute_path)
        os.makedirs(directory, exist_ok=True)

        try:
            fd, tmp_path = tempfile.mkstemp(
                prefix=".dashboard_tmp_", suffix=".json", dir=directory,
            )
            try:
                with os.fdopen(fd, "w", encoding="utf-8") as f:
                    json.dump(spec, f, indent=2)
                os.replace(tmp_path, absolute_path)
            except Exception:
                # Clean up the temp file if the rename failed.
                try:
                    os.remove(tmp_path)
                except OSError:
                    pass
                raise
        except OSError as e:
            logger.warning("Failed to write dashboard %s: %s", absolute_path, e)
            return False
        return True

    # ------------------------------------------------------------------
    # ID generation (must match C++ slugify / uniqueUserIdFromLabel exactly)
    # ------------------------------------------------------------------

    @staticmethod
    def _slugify(label: str) -> str:
        s = (label or "").lower()
        s = re.sub(r"[^a-z0-9]+", "-", s)
        s = s.strip("-")
        if not s:
            s = "dashboard"
        return s

    def _unique_user_id_from_label(self, label: str) -> str:
        base = self._slugify(label)
        taken = self._existing_ids()
        if base not in taken:
            return base
        for i in range(2, 1000):
            candidate = f"{base}-{i}"
            if candidate not in taken:
                return candidate
        # Astronomically unlikely; last resort.
        return f"{base}-{int(time.time() * 1000)}"

File: src/managers/dashboardmanager.cpp
// dashboardmanager.cpp

#include "dashboardmanager.h"

#include <QDateTime>
#include <QDir>
#include <QFile>
#include <QFileInfo>
#include <QJsonDocument>
#include <QJsonObject>
#include <QLoggingCategory>
#include <QRegularExpression>
#include <QTemporaryFile>

Q_LOGGING_CATEGORY(lcDashboards, "octave.dashboards")

DashboardManager::DashboardManager(QObject *parent)
    : QObject(parent)
{
}

void DashboardManager::setPresetsDir(const QString &absolutePath)
{
    m_presetsDir = absolutePath;
}

void DashboardManager::setUserDir(const QString &absolutePath)
{
    m_userDir = absolutePath;
    rescanAll();
}

void DashboardManager::refresh()
{
    rescanAll();
}

// ---------------------------------------------------------------------------
// Lookups
// ---------------------------------------------------------------------------

QVariantMap DashboardManager::findEntry(const QString &id) const
{
    for (const QVariant &v : m_dashboards) {
        QVariantMap m = v.toMap();
        if (m.value(QStringLiteral("id")).toString() == id)
            return m;
    }
    return {};
}

QSet<QString> DashboardManager::existingIds() const
{
    QSet<QString> ids;
    for (const QVariant &v : m_dashboards)
        ids.insert(v.toMap().value(QStringLiteral("id")).toString());
    return ids;
}

bool DashboardManager::isBuiltIn(const QString &id) const
{
    return findEntry(id).value(QStringLiteral("builtIn")).toBool();
}

QVariant DashboardManager::loadDashboard(const QString &id)
{
    QVariantMap entry = findEntry(id);
    if (entry.isEmpty()) {
        qCWarning(lcDashboards) << "loadDashboard: unknown id" << id;
        return QVariantMap{};
    }
    return readFullSpec(entry.value(QStringLiteral("path")).toString());
}

// ---------------------------------------------------------------------------
// Scanning
// ---------------------------------------------------------------------------

void DashboardManager::ensureUserDir() const
{
    if (m_userDir.isEmpty()) return;
    QDir().mkpath(m_userDir);
}

void DashboardManager::rescanAll()
{
    m_dashboards.clear();
    QSet<QString> takenIds;

    // Built-in presets first — these are authoritative; any user JSON that
    // tries to reuse a built-in id gets rejected below.
    if (!m_presetsDir.isEmpty()) {
        QDir d(m_presetsDir);
        if (d.exists()) {
            const auto files = d.entryInfoList(
                {QStringLiteral("*.json")}, QDir::Files, QDir::Name);
            for (const QFileInfo &fi : files) {
                QVariantMap header = readHeader(fi.absoluteFilePath());
                if (header.isEmpty()) continue;
                QString id = header.value(QStringLiteral("id")).toString();
                if (id.isEmpty() || takenIds.contains(id)) {
                    qCWarning(lcDashboards)
                        << "Skipping preset with missing or duplicate id:"
                        << fi.fileName();
                    continue;
                }
                header.insert(QStringLiteral("builtIn"), true);
                header.insert(QStringLiteral("path"), fi.absoluteFilePath());
                m_dashboards.append(header);
                takenIds.insert(id);
            }
        } else {
            qCWarning(lcDashboards) << "Presets dir does not exist:" << m_presetsDir;
        }
    }

    // User dashboards.
    ensureUserDir();
    if (!m_userDir.isEmpty()) {
        QDir u(m_userDir);
        if (u.exists()) {
            const auto files = u.entryInfoList(
                {QStringLiteral("*.json")}, QDir::Files, QDir::Name);
            for (const QFileInfo &fi : files) {
                QVariantMap header = readHeader(fi.absoluteFilePath());
                if (header.isEmpty()) continue;
                QString id = header.value(QStringLiteral("id")).toString();
                if (id.isEmpty()) {
                    qCWarning(lcDashboards)
                        << "Skipping user dashboard with missing id:"
                        << fi.fileName();
                    continue;
                }
                if (takenIds.contains(id)) {
                    qCWarning(lcDashboards)
                        << "Skipping user dashboard — id collides with built-in:"
                        << id << "(" << fi.fileName() << ")";
                    continue;
                }
                header.insert(QStringLiteral("builtIn"), false);
                header.insert(QStringLiteral("path"), fi.absoluteFilePath());
                m_dashboards.append(header);
                takenIds.insert(id);
            }
        }
    }

    qCInfo(lcDashboards) << "Scanned dashboards:"
                         << m_dashboards.size() << "total";
    emit dashboardsChanged();
}

// ---------------------------------------------------------------------------
// JSON I/O
// ---------------------------------------------------------------------------

QVariantMap DashboardManager::readHeader(const QString &absolutePath) const
{
    QVariantMap full = readFullSpec(absolutePath);
    if (full.isEmpty()) return {};

    QVariantMap header;
    header.insert(QStringLiteral("id"),    full.value(QStringLiteral("id")));
    header.insert(QStringLiteral("label"), full.value(QStringLiteral("label")));
    return header;
}

QVariantMap DashboardManager::readFullSpec(const QString &absolutePath) const
{
    QFile f(absolutePath);
    if (!f.open(QIODevice::ReadOnly | QIODevice::Text)) {
        qCWarning(lcDashboards) << "Cannot open dashboard file:" << absolutePath;
        return {};
    }
    const QByteArray data = f.readAll();
    f.close();

    QJsonParseError err;
    const QJsonDocument doc = QJsonDocument::fromJson(data, &err);
    if (err.error != QJsonParseError::NoError) {
        qCWarning(lcDashboards) << "Invalid JSON in" << absolutePath
                                << ":" << err.errorString();
        return {};
    }
    if (!doc.isObject()) {
        qCWarning(lcDashboards) << "Dashboard JSON root must be an object:"
                                << absolutePath;
        return {};
    }
    return doc.object().toVariantMap();
}

bool DashboardManager::writeSpec(const QString &absolutePath, const QVariantMap &spec)
{
    QJsonDocument doc(QJsonObject::fromVariantMap(spec));

    QFileInfo fi(absolutePath);
    QDir().mkpath(fi.absolutePath());

    // Atomic write: temp file + rename.
    QTemporaryFile tmp(fi.absolutePath() + QStringLiteral("/.dashboard_tmp_XXXXXX.json"));
    tmp.setAutoRemove(false);
    if (!tmp.open()) {
        qCWarning(lcDashboards) << "Cannot create temp file:" << tmp.errorString();
        return false;
    }
    tmp.write(doc.toJson(QJsonDocument::Indented));
    const QString tmpName = tmp.fileName();
    tmp.close();

#ifdef Q_OS_WIN
    if (QFile::exists(absolutePath)) QFile::remove(absolutePath);
#endif
    if (!QFile::rename(tmpName, absolutePath)) {
        qCWarning(lcDashboards) << "Atomic rename failed, falling back to direct write";
        QFile::remove(tmpName);
        QFile direct(absolutePath);
        if (!direct.open(QIODevice::WriteOnly | QIODevice::Text)) {
            qCWarning(lcDashboards) << "Cannot open" << absolutePath << "for writing";
            return false;
        }
        direct.write(doc.toJson(QJsonDocument::Indented));
        direct.close();
    }
    return true;
}

// ---------------------------------------------------------------------------
// Mutations
// ---------------------------------------------------------------------------

QString DashboardManager::saveDashboard(const QVariantMap &spec)
{
    const QString id    = spec.value(QStringLiteral("id")).toString();
    const QString label = spec.value(QStringLiteral("label")).toString();

    if (id.isEmpty() || label.isEmpty()) {
        qCWarning(lcDashboards) << "saveDashboard requires id and label";
        return {};
    }
    if (m_userDir.isEmpty()) {
        qCWarning(lcDashboards) << "User dashboards dir not configured";
        return {};
    }

    // Don't allow saving over a built-in id.
    QVariantMap existing = findEntry(id);
    if (!existing.isEmpty() && existing.value(QStringLiteral("builtIn")).toBool()) {
        qCWarning(lcDashboards)
            << "Cannot save: id" << id << "collides with a built-in preset";
        return {};
    }

    const QString path = m_userDir + QDir::separator() + id + QStringLiteral(".json");
    if (!writeSpec(path, spec)) return {};

    rescanAll();
    return id;
}

bool DashboardManager::deleteDashboard(const QString &id)
{
    QVariantMap entry = findEntry(id);
    if (entry.isEmpty()) return false;
    if (entry.value(QStringLiteral("builtIn")).toBool()) {
        qCWarning(lcDashboards) << "Refusing to delete built-in dashboard:" << id;
        return false;
    }

    const QString path = entry.value(QStringLiteral("path")).toString();
    if (!QFile::remove(path)) {
        qCWarning(lcDashboards) << "Failed to delete:" << path;
        return false;
    }

    rescanAll();
    return true;
}

QString DashboardManager::duplicateDashboard(const QString &sourceId,
                                             const QString &newLabel)
{
    QVariantMap entry = findEntry(sourceId);
    if (entry.isEmpty()) {
        qCWarning(lcDashboards) << "duplicateDashboard: unknown source id" << sourceId;
        return {};
    }

    QVariantMap spec = readFullSpec(entry.value(QStringLiteral("path")).toString());
    if (spec.isEmpty()) return {};

    const QString finalLabel = newLabel.isEmpty()
        ? (entry.value(QStringLiteral("label")).toString() + QStringLiteral(" (Copy)"))
        : newLabel;
    const QString newId = uniqueUserIdFromLabel(finalLabel);

    spec.insert(QStringLiteral("id"),    newId);
    spec.insert(QStringLiteral("label"), finalLabel);

    return saveDashboard(spec);
}

// ---------------------------------------------------------------------------
// ID generation
// ---------------------------------------------------------------------------

QString DashboardManager::slugify(const QString &label)
{
    QString s = label.toLower();
    s.replace(QRegularExpression(QStringLiteral("[^a-z0-9]+")), QStringLiteral("-"));
    while (s.startsWith('-')) s.remove(0, 1);
    while (s.endsWith('-'))   s.chop(1);
    if (s.isEmpty()) s = QStringLiteral("dashboard");
    return s;
}

QString DashboardManager::uniqueUserIdFromLabel(const QString &label) const
{
    const QString base = slugify(label);
    const QSet<QString> taken = existingIds();

    if (!taken.contains(base)) return base;

    for (int i = 2; i < 1000; ++i) {
        const QString candidate = base + QStringLiteral("-") + QString::number(i);
        if (!taken.contains(candidate)) return candidate;
    }
    // Astronomically unlikely; last resort.
    return base + QStringLiteral("-") + QString::number(QDateTime::currentMSecsSinceEpoch());
}
